### PR TITLE
Support non-explicit output argument calls to Numpy ufunc's with NRT.

### PR DESCRIPTION
**Pull request by jriehl**
This PR builds on PR #1077 (sklam/numba_runtime), and also requires PR #1078.

This PR enables the following example to work:
```python
from numba import *
import numpy as np

@njit(nrt=True)
def myadd(a0, a1):
    return np.add(a0, a1)

x, y = np.linspace(0,1,10), np.linspace(1,2,10)
print(myadd(x, y))
```

### CODE REVIEW

**Comment by seibert** [comment]
Now that NRT is merged, this PR needs to remerge master again, I think.

**Comment by seibert** [comment]
Can you add some test cases for Fortran-ordered arrays?  After a few experiments with Numpy, it seems that operations where all the arguments are in Fortran-order result in a Fortran-order output.  If any argument is in C-order, then the output is in C-order.

**Comment by jriehl** [comment]
I added one unit test, and found us lacking.  Fix is in.  More required, or is one enough?

**Comment by seibert** [comment]
That test looks like it covers the two important cases.

File: numba/tests/test_ufuncs.py [hunk 1]
```diff
@@ -1014,6 +1017,65 @@ def myadd(a0, a1):
...
+            input1_type = types.Array(types.uint64, x.ndim, 'C')
+            input2_type = types.Array(types.uint64, y.ndim, 'C')
+
+            cr = self.cache.compile(pyfunc, (input1_type, input2_type),
+                                    flags=enable_nrt_flags)
+            cfunc = cr.entry_point
+
+            expected = np.add(x, y)
+            result = cfunc(x, y)
+            self.assertTrue(np.all(result == expected))
```
Comment (pitrou) [inline numba/tests/test_ufuncs.py]: It seems Numpy's `==` operator implementation doesn't check array shapes. Perhaps use `np.testing.assert_array_equal` instead?

File: numba/typing/npydecl.py [hunk 2]
```diff
@@ -62,13 +62,27 @@ def _handle_inputs(cls, ufunc, args, kws):
...
+        layout = None
+        if ndims > 0:
+            layout = 'C'
+            layouts = [x.layout if isinstance(x, types.Array) else ''
+                       for x in args]
+            if 'C' not in layouts:
+                if 'F' in layouts:
+                    layout = 'F'
+                elif 'A' in layouts:
+                    layout = 'A'
```
Comment (pitrou) [inline numba/typing/npydecl.py]: Just for the record, output layout 'A' isn't implemented, right?
Perhaps this should simply be a TypingError?

File: numba/tests/test_ufuncs.py (hunk 1 above)
Comment (seibert) [inline numba/tests/test_ufuncs.py]: Agreed.  assert_array_equal() is the better idiom to use.

File: numba/typing/npydecl.py (hunk 2 above)
Comment (jriehl) [inline numba/typing/npydecl.py]: You're right.  One of the major points of type checking is to error out early if bad code or no code will/can be generated.  Fix incoming.

File: numba/tests/test_ufuncs.py (hunk 1 above)
Comment (jriehl) [inline numba/tests/test_ufuncs.py]: Fix incoming.

**Pull request merged**